Advanced Indexing Part 1 -- Purely Integer Array Indexing

killeent opened the pull request:
This PR implements "Purely Integer Array" advanced indexing semantics in PyTorch, from NumPy, as documented here: https://docs.scipy.org/doc/numpy/reference/arrays.indexing.html#purely-integer-array-indexing. It should work with gets/sets and autograd, and allows for LongTensor indexers as well.

```
 x = torch.arange(0, 25).view(5, 5)

  0   1   2   3   4
  5   6   7   8   9
 10  11  12  13  14
 15  16  17  18  19
 20  21  22  23  24
[torch.FloatTensor of size 5x5]

x[[0,4,3], [1,3,0]]

  1
 23
 15
[torch.FloatTensor of size 3]
```

Begins to address #1080.

killeent committed:
temp commit advanced indexing

killeent committed:
basic advanced indexing with ndim 1d long tensors

killeent committed:
support arbitrary indexing shapes for output

fmassa commented on torch/csrc/generic/Tensor.cpp:
```diff
@@ -498,14 +498,115 @@ static bool THPTensor_(_indexOnce)(PyObject *index, int &indexed_dim,
+
+  THIndexTensor *linearIndices = THIndexTensor_(newWithSize1d)(LIBRARY_STATE THIndexTensor_(nElement)(LIBRARY_STATE first));
+  for (ptrdiff_t i = 0; i < THIndexTensor_(nElement)(LIBRARY_STATE linearIndices); ++i) {
+    long stride = 1; // is this an incorrect assumption? could the stride at the last dim be non-one?
```
If the tensor is not contiguous, then the stride of the last dimension might not be 1.

killeent replied:
@fmassa yeah this has been fixed locally to use the tensor strides directly!

killeent committed:
arbitrary sequence indexers

killeent committed:
use strides directly for linear index calculation

killeent committed:
tests in passing state minus broadcasting

killeent committed:
clean up checkAdvancedIndexing

killeent committed:
clean up code, add memory frees/decrefs

killeent committed:
support empty indexer

killeent committed:
hacky commit broadcast semantics

killeent committed:
support transposed tensors

killeent committed:
move advanced indexing into getvalue

killeent committed:
refactor out argument conversion code into separate method

killeent committed:
refactor out linear index calculation

killeent committed:
advanced index set + unit tests

killeent committed:
fix minor mem leak

killeent committed:
check available at interpreter level

killeent committed:
get advanced indexing autograd working for non duplicates, unit test for both cases (second fails)

killeent committed:
advancedIndexAdd, pass tests for autograd for indexing duplicates

killeent committed:
refactor to share init/cleanup code among advanced indexing ops

killeent committed:
minor fixes to advanced index add

killeent committed:
advanced indexing works w subsets of Tensors that arent contiguous in memory

killeent committed:
advanced indexing big test

killeent committed:
changes necessary from rebase: update expandNd

killeent commented:
This is now in a reviewable state cc @fmassa.

fmassa commented on torch/csrc/generic/Tensor.cpp:
```diff
@@ -496,14 +503,354 @@ static bool THPTensor_(_indexOnce)(PyObject *index, int &indexed_dim,
+      // necessary
+
+      THTensor *contiguous;
+      if (THTensor_(isContiguous(LIBRARY_STATE ((THPTensor*)src)->cdata))) {
```
You can just call `THTensor_(newContiguous)`, as [it internally checks if the tensor is contiguous](https://github.com/pytorch/pytorch/blob/master/torch/lib/TH/generic/THTensor.c#L190-L199) before performing the memory copy (skipping if it is contiguous).

fmassa commented on torch/csrc/generic/Tensor.cpp:
```diff
@@ -496,14 +503,354 @@ static bool THPTensor_(_indexOnce)(PyObject *index, int &indexed_dim,
+  if (success) {
+    // Verify src tensor is contiguous before flattening
+    THTensor *contiguous;
+    if (THTensor_(isContiguous(LIBRARY_STATE src))) {
```
Same thing here about the `contiguous` check.

killeent committed:
call newcontiguous where appropriate

killeent committed:
index bounds checking

kentsommer commented:
Is this indexing supposed to work on tensors that have been pushed to the gpu with ```.cuda()```?

If it is, I encountered this issue today while testing the indexing out:
```
(apytorch) kent@kbox ~/Documents/misc-git-repos $ python
Python 2.7.12 (default, Nov 19 2016, 06:48:10) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> import numpy as np
>>> x = torch.arange(0,80).view(1,10,8)
>>> x[[0], [1, 2, 3, 4, 5, 6, 7, 8, 9], [2]]

 10
 18
 26
 34
 42
 50
 58
 66
 74
[torch.FloatTensor of size 9]

>>> y = x.cuda()
>>> y[[0], [1, 2, 3, 4, 5, 6, 7, 8, 9], [2]]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: 208 is not a device at /home/kent/Documents/misc-git-repos/pytorch/torch/lib/THC/THCGeneral.c:206
>>> y[[0], [1, 2, 3, 4, 5, 6, 7, 8, 9], [2]]
THCudaCheck FAIL file=/home/kent/Documents/misc-git-repos/pytorch/torch/lib/THC/generic/THCStorage.c line=32 error=77 : an illegal memory access was encountered
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: cuda runtime error (77) : an illegal memory access was encountered at /home/kent/Documents/misc-git-repos/pytorch/torch/lib/THC/generic/THCStorage.c:32
>>>
```

killeent commented:
@kentsommer Yes in theory it should work for CUDA. Let me take a look.

fmassa reviewed (changes requested):
I did a first pass and it looks good for the most part.
I think there is a problem with the autograd part that would be good to have fixed (in the worst case, you can have a check if the `grad_output` is `volatile` (meaning no grad of grad), and raise an error if not.
Also, for the more general case `x[:, [1, 2], [2]]`, the linear indices might grow quite a bit, but we would need to check to see if that would have an impact or not.

killeent commented on torch/csrc/generic/Tensor.cpp:
```diff
@@ -496,14 +503,363 @@ static bool THPTensor_(_indexOnce)(PyObject *index, int &indexed_dim,
+      THIndexTensor *oned = THIndexTensor_(newView)(LIBRARY_STATE indexer, indexerSize);
+
+      // Actually laod the value, and verify it is not out-of-bounds
+      long indexAtDim = THIndexTensor_(get1d)(LIBRARY_STATE oned, i);
```
In the short term to get this checked in, I may have to keep it all on the GPU. Unfortunately dealing with non-generic types in `generic/Tensor.cpp` is a little tricky. Ultimately, all the index calculation will be done on the GPU, and only once we are ready to do the indexSelect will we copy that indexing Tensor back to the GPU.

In the long term, index calculation could be a kernel in THC.

killeent committed:
make work with cuda

killeent commented:
@kentsommer This should now work with CUDA. The issue was that we were trying to use CPU Tensors to index into the CUDA Tensor. Now we copy the indices onto the GPU before calling the `index_*` functions. cc @fmassa this means that all index calculation is being done on the CPU :).

I will take a look at grad of grad tomorrow.

killeent committed:
make adv indexing work with variable, gradgrad

killeent commented:
Okay, I have now moved the `advanced_index_add` call into Variable. This required adding a new function to _functions/tensor.py` that handles forwards and backwards for this method, including adding a new function `advanced_index_select` for the backward's implementation. I don't understand enough about autograd to have confidence in this implementation, but it passes the minimal tests I have added.

apaszke commented on torch/autograd/_functions/tensor.py:
```diff
@@ -14,13 +14,19 @@ def forward(ctx, i, index):
+        if ctx.advanced_indexing:
+            grad_input._advanced_index_add(ctx.index, grad_output)
+        else:
+            grad_input[ctx.index] = grad_output
```
This path is taken only when `ctx.index` is at least one number or slice, right? No tensors/lists

killeent replied:
Advanced indexing is only triggered if we have `ndim` sequence objects, where `ndim` is the number of dimensions of the Tensor being indexed. In subsequent PRs, with more advanced indexing functionality, this will be relaxed.

apaszke replied:
I see. Thanks

apaszke commented on test/test_torch.py:
```diff
@@ -2496,6 +2496,247 @@ def test_index(self):
+        def consec(size, start=1):
+            sequence = torch.ones(int(torch.Tensor(size).prod(0)[0])).cumsum(0)
+            sequence.add_(start - 1)
+            return sequence.resize_(*size)
```
Never use `resize_` like that (tensor content after resize is unspecified). This is what `view` is for. It works now, but would silently give incorrect results if the tensor was noncontiguous

killeent replied:
Will change.

apaszke reviewed (commented):
I glanced over autograd and it looks ok to me

killeent committed:
use .view() instead of .resize() in indexing test

killeent committed:
smaller tensors for advancedindex_big, run in cuda as well

soumith commented on torch/autograd/_functions/tensor.py:
```diff
@@ -195,6 +201,29 @@ def backward(ctx, grad_output):
+        assert not ctx.needs_input_grad[1]
+        if ctx.needs_input_grad[2]:
+            ctx.adv_index = adv_index
+        ctx.mark_dirty(tensor1)
```
i dont think you need to mark_dirty tensor1, as it's not an in-place or shared-storage operation (regular indexing is)

soumith replied:
oh nevermind it's inplace

soumith commented:
this is now merged into master!

vadimkantorov commented:
Just updated my PyTorch. `x[[0,4,3], [1,3,0]]` from the OP works, but `x[[0,4,3]]` doesn't. Is it ok?

`x[[0,4,3], :]` works as well as `x[torch.LongTensor([0,4,3])]`.

soumith commented:
yes it's a todo in the code. will be fixed in a next PR

ngimel commented:
@killeent, I have a cuda test failing when I build from master, ```python test_cuda.py TestCuda.test_advancedindex```. It's failing in assert_set_eq when the indexer has dupes, and I've added some printouts there:
```
tensor
  0   1   2   3   4
  5   6   7   8   9
 10  11  12  13  14
 15  16  17  18  19
indexer [slice(None, None, None), [0, 1, 1, 2, 2]]
val
 16  12   0  14  10
 19  17   2   7   4
  5   9   6   1  15
 13  11   3   8  18
[torch.cuda.DoubleTensor of size 4x5 (GPU 0)]

pyt 
 16  12  14   3   4
 19  17   7   8   9
  5   9   1  13  14
 13  11   8  18  19
[torch.cuda.DoubleTensor of size 4x5 (GPU 0)]
 numt 
 16   0  10   3   4
 19   2   4   8   9
  5   6  15  13  14
 13   3  18  18  19
```
Apparently, in the correct output (numt) the second of the duped columns should be copied to tensor from val, and in my failing case (pyt) the first one is. Can you point me to where in the backend this is happening? It's passing on most architectures, so that's not an outright bug, but failing on one.

vadimkantorov commented:
In addition to `x[[0,4,3]]`, `torch.arange(0, 25)[[0,4,3]]` doesn't work either, except the workaround `torch.arange(0, 25)[[0,4,3], :]` isn't possible for the lack of dimensions. (I guess this one may exist on TODO as well, but since stumbled upon it, will leave it here for completeness).
